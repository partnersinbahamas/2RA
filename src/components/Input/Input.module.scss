.primary {
  [id="heading"] {
    margin: 3px;
    color: #1ea7fd;
  }
  [id="label"] {
    background-color: #D6EFFF;
    border-radius: .3rem;
  }
  [id="input"] {
    border-color: #1ea7fd;
    color: #1ea7fd;
    font-weight: 700;

    &:focus {
      border-color: #0085D8;
    }
  }
  [id="error-message"] {
    color: #FF9058
  }
  &-error {
    border-color: #FF9058 !important;
  }
}
