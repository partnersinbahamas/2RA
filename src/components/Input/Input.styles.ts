import styled from "styled-components";

import { COOL_GRAY_400, COOL_GRAY_500, RED_400,  } from "../Colors/colors";

const InputWrapper = styled.div`
  width: 100%;
  font-size: 0.75rem;

  display: flex;
  font-family: 'Poppins', sans-serif;
  flex-direction: column;
`;
const InputComponent = styled.input`
  width: 100%;
  padding: 0.625rem;
  box-sizing: border-box;
  font-family: 'Poppins', sans-serif;
  border: 2px solid ${COOL_GRAY_400};
  border-radius: 0;
  &.error {
    border-color: ${RED_400};
  }
  &:focus {
    outline: none;
    border-color: ${COOL_GRAY_500};
  }
`;
const Label = styled.label`
  margin: 0;
  margin-bottom: 5px;
  font-weight: 600;
  font-size: 0.625rem;
`;
const Error = styled.span`
  color: ${RED_400};
`;
const Heading = styled.h3`
  margin: 0;
`
export{ InputWrapper, InputComponent, Label, Error, Heading };