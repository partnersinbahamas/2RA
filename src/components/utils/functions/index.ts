export const wait = (ms: number) =>
  new Promise(resolve => setTimeout(resolve, ms));

export function isDislayCountFormatValid(format: string) {
  const regex = /^X{1,16}$/;
  return regex.test(format);
}

export function numberFixTo(number: number | string, format: string): string {
  if (format && !isDislayCountFormatValid(format)) {
    throw new Error(
      'fixTo: fromat invalid. Please enter the correct format: X*16. Example: XXX.',
    );
  }

  return Number(number).toFixed(format?.length);
}
