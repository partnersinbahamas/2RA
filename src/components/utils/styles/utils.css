.m-0 {
  margin: 0 !important;
}

/* top */
.mt-auto {
  margin-top: auto !important;
}

.mt-0 {
  margin-top: 0 !important;
}

.mt-1-2 {
  margin-top: 0.5rem !important;
}

.mt-1-4 {
  margin-top: 0.25rem !important;
}

.mt-1 {
  margin-top: 1rem !important;
}

.mt-2 {
  margin-top: 2rem !important;
}

/* right */
.mr-auto {
  margin-right: auto !important;
}

.mr-0 {
  margin-right: 0 !important;
}

.mr-1-2 {
  margin-right: 0.5rem !important;
}

.mr-1-4 {
  margin-right: 0.25rem !important;
}

.mr-1 {
  margin-right: 1rem !important;
}

.mr-2 {
  margin-right: 2rem !important;
}

/* bottom */
.mb-auto {
  margin-bottom: auto !important;
}

.mb-0 {
  margin-bottom: 0 !important;
}

.mb-1-2 {
  margin-bottom: 0.5rem !important;
}

.mb-1-4 {
  margin-bottom: 0.25rem !important;
}

.mb-1 {
  margin-bottom: 1rem !important;
}

.mb-2 {
  margin-bottom: 2rem !important;
}

/* left */
.ml-auto {
  margin-left: auto !important;
}

.ml-0 {
  margin-left: 0 !important;
}

.ml-1-2 {
  margin-left: 0.5rem !important;
}

.ml-1-4 {
  margin-left: 0.25rem !important;
}

.ml-1 {
  margin-left: 1rem !important;
}

.ml-2 {
  margin-left: 2rem !important;
}
