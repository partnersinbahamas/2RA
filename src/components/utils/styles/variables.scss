$breakpoint-mobile: 576px; // minimal width for mobile
$breakpoint-tablet: 768px; // minimal width for tablet
$breakpoint-desktop: 1200px; // minimal width for desktop
