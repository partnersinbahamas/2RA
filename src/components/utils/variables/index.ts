import { TStile } from '../types/types';
export const stiles: TStile[] = ['mute', 'default', 'primary'];
export const levels: TLevel[] = [1, 2, 3, 4, 5, 6];
