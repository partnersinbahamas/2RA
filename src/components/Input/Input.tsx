import React, { InputHTMLAttributes, useCallback } from 'react';
import classNames from 'classnames';
import { Error, InputComponent, InputWrapper, Label, Heading } from './Input.styles';
import styles from './Input.module.scss';
import { TStile } from '../utils/types/types';
import defaultProps from '../utils/variables/defaultProps';

export interface TProps extends InputHTMLAttributes<HTMLInputElement> {
  title?: string;
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onChangeValue?: (value: string) => void;
  error?: string;
  stile?: TStile;
  disabled?: boolean;
}

export const Input: React.FC<TProps> = ({
  title,
  value,
  onChange,
  onChangeValue,
  error,
  stile = defaultProps.stile,
  disabled,
  ...props
}) => {
  const { className, ...inputProps } = props;
  const label = title ? `${title}-label` : 'label';

  const handleChange = useCallback(
    (event: React.ChangeEvent<HTMLInputElement>) => {
      const { value } = event.target;
      if (onChange) onChange(event);
      if (onChangeValue) onChangeValue(value);
    },
    [onChange, onChangeValue],
  );

  return (
    <InputWrapper disabled={disabled} className={classNames(styles[stile], className)}>
      <Label
        data-name="label"
        aria-label={label}
        stile={stile}
        disabled={disabled}
        className={`${className}-label`}
      >
        <Heading data-name="heading" stile={stile} className={`${className}-heading`}>
          {title}
        </Heading>
        <InputComponent
          data-name="input"
          data-error={!!error}
          stile={stile}
          value={value}
          disabled={disabled}
          {...inputProps}
          onChange={handleChange}
          className={classNames({ error }, { [`${className}-error`]: error }, `${className}-input`)}
        />
      </Label>
      {error && (
        <Error data-name="error-message" className={`${className}-error-message`}>
          {error}
        </Error>
      )}
    </InputWrapper>
  );
};
