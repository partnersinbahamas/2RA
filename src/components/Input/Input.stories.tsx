import { Meta, StoryObj } from '@storybook/react';
import { within, userEvent } from '@storybook/test';
import { story } from '../utils/variables/story';

import { ConnectOnChange } from '../../../.storybook/decorators';
import { Input } from './Input';

const meta = {
  title: 'Athomic/Input',
  component: Input,
  tags: ['autodocs'],
  argTypes: {
    title: story.title,
    stile: story.stile,
    value: story.stringValue,
    error: story.stringError,
    disabled: story.disabled,
    type: story.inputType,
    onChange: story.onChange,
    className: story.className,
  },
  parameters: {
    docs: {
      description: {
        component: `
Please to style your custom input component use the mockup bellow.
### Guide styles

    .className {
      [data-name="heading"] {}
      [data-name="label"] {}
      [data-name="input"] {}
      [data-name="error-message"] {}
      
      &-error {}
    }
`,
      },
    },
  },
  decorators: [ConnectOnChange],
} satisfies Meta<typeof Input>;
export default meta;
type Story = StoryObj<typeof meta>;
export const Default: Story = {
  args: {
    title: 'Input',
    value: 'value',
  },
};

export const Primary: Story = {
  args: {
    stile: 'primary',
    value: 'Primary',
    title: 'Input Primary',
  },
};

export const Error: Story = {
  args: {
    title: 'Handled error',
    error: '...pzzz..pz',
  },
};
export const Play: Story = {
  args: {
    title: 'Play function',
  },
  play: async context => {
    const canvas = within(context.canvasElement);

    const inputElement = canvas.getByRole('textbox');
    await userEvent.type(inputElement, ' play fn', { delay: 100 });
  },
};
