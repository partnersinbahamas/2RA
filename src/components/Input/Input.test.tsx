import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { stiles } from '../utils/variables';
import { Input } from './Input';

describe('Input', () => {
  const user = userEvent.setup();

  it('renders correctly', () => {
    const { container } = render(<Input />);
    expect(container).toMatchSnapshot();
  });
  it('renders with input typing', async () => {
    render(<Input />);

    const inputElement = screen.getByRole('textbox');
    await user.type(inputElement, 'user typing...');
    expect(inputElement).toHaveValue('user typing...');
  });
  it('renders with error', () => {
    const { container } = render(<Input error="error" />);

    const errorElement = screen.queryByText('error');
    expect(errorElement).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
  it.each(stiles)('renders with each stile', stile => {
    const title = 'Input';
    const { container } = render(<Input stile={stile} title={title} />);

    const labelElement = screen.getByLabelText(`${title}-label`);
    const headingElement = screen.getByRole('heading', { name: title });
    const inputElement = screen.getByRole('textbox');

    expect(labelElement.getAttribute('stile')).toBe(stile);
    expect(headingElement.getAttribute('stile')).toBe(stile);
    expect(inputElement.getAttribute('stile')).toBe(stile);

    expect(container).toMatchSnapshot();
  });
});
