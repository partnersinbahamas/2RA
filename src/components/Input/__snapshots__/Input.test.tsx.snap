// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders correctly 1`] = `
<div>
  <div
    class="sc-beySPh fOhtuL default"
  >
    <label
      aria-label="label"
      class="sc-dmyCSP ihRrHi undefined-label"
      data-name="label"
      stile="default"
    >
      <h3
        class="sc-eDLKkx gVYdGR undefined-heading"
        data-name="heading"
        stile="default"
      />
      <input
        class="sc-guDLey fpCSGi undefined-input"
        data-error="false"
        data-name="input"
        stile="default"
      />
    </label>
  </div>
</div>
`;

exports[`Input renders with each stile 1`] = `
<div>
  <div
    class="sc-beySPh fOhtuL mute"
  >
    <label
      aria-label="Input-label"
      class="sc-dmyCSP eBqJxY undefined-label"
      data-name="label"
      stile="mute"
    >
      <h3
        class="sc-eDLKkx gySIDl undefined-heading"
        data-name="heading"
        stile="mute"
      >
        Input
      </h3>
      <input
        class="sc-guDLey jofGtu undefined-input"
        data-error="false"
        data-name="input"
        stile="mute"
      />
    </label>
  </div>
</div>
`;

exports[`Input renders with each stile 2`] = `
<div>
  <div
    class="sc-beySPh fOhtuL default"
  >
    <label
      aria-label="Input-label"
      class="sc-dmyCSP ihRrHi undefined-label"
      data-name="label"
      stile="default"
    >
      <h3
        class="sc-eDLKkx gVYdGR undefined-heading"
        data-name="heading"
        stile="default"
      >
        Input
      </h3>
      <input
        class="sc-guDLey fpCSGi undefined-input"
        data-error="false"
        data-name="input"
        stile="default"
      />
    </label>
  </div>
</div>
`;

exports[`Input renders with each stile 3`] = `
<div>
  <div
    class="sc-beySPh fOhtuL primary"
  >
    <label
      aria-label="Input-label"
      class="sc-dmyCSP ihRrHi undefined-label"
      data-name="label"
      stile="primary"
    >
      <h3
        class="sc-eDLKkx gVYdGR undefined-heading"
        data-name="heading"
        stile="primary"
      >
        Input
      </h3>
      <input
        class="sc-guDLey fpCSGi undefined-input"
        data-error="false"
        data-name="input"
        stile="primary"
      />
    </label>
  </div>
</div>
`;

exports[`Input renders with error 1`] = `
<div>
  <div
    class="sc-beySPh fOhtuL default"
  >
    <label
      aria-label="label"
      class="sc-dmyCSP ihRrHi undefined-label"
      data-name="label"
      stile="default"
    >
      <h3
        class="sc-eDLKkx gVYdGR undefined-heading"
        data-name="heading"
        stile="default"
      />
      <input
        class="sc-guDLey fpCSGi error undefined-error undefined-input"
        data-error="true"
        data-name="input"
        stile="default"
      />
    </label>
    <span
      class="sc-hLQSwg gpNhwr undefined-error-message"
      data-name="error-message"
    >
      error
    </span>
  </div>
</div>
`;
